Use relative time description for posting timestamps and periodically update these to keep then up-to-date

hdurer opened the pull request.

hdurer committed:
Show the posting date in a more readable, relative to current time format.

This is still static, i.e. doesn't update as time progresses.

hdurer committed:
Add periodic updating of the timestamp displays.

For each buffer we add a timer that periodically checks all timestamps and updates them as needed.
The logic tries to be smart and
- only schedule an update when at least one timestamps display needs changing (although at the moment we update at least every 5 minutes), and
- only do a limited amount of work in each timer callback so as to not block Emacs's interactive work.

hdurer committed:
More robust bad url detection.

alexjgriffith commented on lisp/mastodon-tl.el:
```diff
@@ -48,11 +48,20 @@
 
 (defvar mastodon-tl--buffer-spec nil
   "A unique identifier and functions for each Mastodon buffer.")
+(make-variable-buffer-local 'mastodon-tl--buffer-spec)
```
Nice clean up

alexjgriffith commented on lisp/mastodon-tl.el:
```diff
@@ -163,7 +237,11 @@ Return value from boosted content if available."
              (mastodon-tl--byline-boosted toot)
              " "
-             (format-time-string mastodon-toot-timestamp-format (date-to-time timestamp))
+             (let ((parsed-time (date-to-time timestamp)))
```
If we don't use `timestamp` outside of defining `parsed-time` could we combine this `let` with the first `let` in this function? i.e., replace 

```elisp
(timestamp (mastodon-tl--field 'created_at toot))
```

with

 ```elisp 
(parsed-time (date-to-time (mastodon-tl--field 'created_at toot)))
```

hdurer replied:
Good point!  Done.

alexjgriffith commented on lisp/mastodon-tl.el:
```diff
@@ -356,14 +525,23 @@ UPDATE-FUNCTION is used to recieve more toots."
-                          ,update-function)))
+                          ,update-function)
+            mastodon-tl--timestamp-update-timer (run-at-time mastodon-tl--timestamp-next-update
+                                                             nil ;; don't repeat
```
Is their an easy way to make the auto update on/off configurable?

hdurer replied:
I added a customizable var to switch it all off.

Note that the 5mins mentioned above is the upper limit for a refresh; we could up that now (I'm fairly confident my next-update calculation is working properly now)...

alexjgriffith commented on test/mastodon-tl-tests.el:
```diff
@@ -1,4 +1,5 @@
 (require 'el-mock)
+(require 'cl-macs)
```
Dependent on `cl-lib`. I'm not sure what the policy is on including `cl-lib` in tests though, will have to be checked by @jdenen

hdurer replied:
I asked in the matrix channel... let's see what he thinks.

I personally don't think it's an issue.
As I understand it, use of the "undesirable" files is even considered palatable as long as they are only required for bytecompiling (i.e. macros are ok if they don't expand into something undesirable to be used at runtime).  Surely the same reasoning would apply for tests: we don't want to burden users to require these files for normal operation but for tests (or compiling) it's ok.

alexjgriffith commented on lisp/mastodon-tl.el:
```diff
@@ -146,6 +155,71 @@ Return value from boosted content if available."
       (cdr (assoc field toot))))
 
+(defun mastodon-tl--relative-time-details (timestamp &optional current-time)
+  "Returns cons of (descriptive string . next change) for the TIMESTAMP.
```
checkdoc suggests using the imperative return rather than returns. However, it looks like their are endemic issues with the docs in this file, so perhaps we can go through and fix all the doc issues in a separate PR.

hdurer replied:
Agreed.  Let's keep it separate.  It's a habbit I picked up at work.  3rd person is the correct Java style here.

alexjgriffith commented on lisp/mastodon-tl.el:
```diff
@@ -146,6 +155,71 @@ Return value from boosted content if available."
+           (t
+            (funcall regular-response seconds-difference (* 60 60 24 365.25) "year")))))
+    (cons (car relative-result)
+          (time-add timestamp (seconds-to-time (cdr relative-result))))))
```
These human readable time strings are the bomb!

alexjgriffith reviewed (commented):
The new time strings look great and work on both my mac and windows machine without an issue. Only one nitpick and a couple suggestions for customizations.

Could we make the update time (5min) a custom-var, or have an on/off custom for running updates in the background?

hdurer committed:
Catch any errors thrown during url-retrieve.

On Emacs24 I've been able to reliably fail url fetching which exposed issues in loading a timeline (it aborts the loading).
This catches any errors, marking the image load as failed so that we won't retry (retries are a TODO item I guess) and then succeeds the function so the rest of the timeline loading can proceed.

hdurer committed:
Review comments from alexjgriffith:

- Simplify let structure in byline generation
- Allow disabling of relative timestamps via customization

Maybe still to do:
- use of cl-macs in tests

Punted to a later PR:
- Using correct form in docstrings throughout

jdenen committed:
Bump version to 0.7.1

alexjgriffith commented on lisp/mastodon-tl.el:
```diff
@@ -163,7 +245,10 @@ Return value from boosted content if available."
+             (propertize
+              (format-time-string mastodon-toot-timestamp-format parsed-time)
+              'timestamp parsed-time
+              'display (mastodon-tl--relative-time-description parsed-time))
```
Could we use the non relative timestamp here when `mastodon-tl--enable-relative-timestamps` is nil?

``` emacs-lisp
    'display (if mastodon-tl--enable-relative-timestamps
                           (mastodon-tl--relative-time-description parsed-time)
                         parsed-time)
```

alexjgriffith committed:
use absolute time stamps when mastodon-tl--enable-relative-timestamps is nil

alexjgriffith reviewed (approved):
Went ahead and made the one small change.

Works fine, relative time strings are amazing. Setting `mastodon-tl--enable-relative-timestamps` to nil now disables the periodic updates and reverts to the absolute strings. We can change this behaviour in the future if desired.
